server/expression/array.go

```diff
@@ -147,163 +143,27 @@ func (array *Array) WithResolvedChildren(children []any) (any, error) {
-	} else {
-		// TODO: we need to remove GMS types from all of our expressions so that we can remove this. For now, we have to
-		// handle all possible GMS types and make any conversions for types that are not supported by Postgres
-		switch paramType.Type() {
-		case query.Type_INT8:
-			*val = int16((*val).(int8))
-			paramType = pgtypes.Int16
-		case query.Type_INT16:
-			paramType = pgtypes.Int16
-		case query.Type_INT24, query.Type_INT32:
-			paramType = pgtypes.Int32
-		case query.Type_INT64:
-			paramType = pgtypes.Int64
-		case query.Type_UINT8:
-			*val = int64((*val).(uint8))
-			paramType = pgtypes.Int64
-		case query.Type_UINT16:
-			*val = int64((*val).(uint16))
-			paramType = pgtypes.Int64
-		case query.Type_UINT24, query.Type_UINT32:
-			*val = int64((*val).(uint32))
-			paramType = pgtypes.Int64
-		case query.Type_UINT64:
-			*val = int64((*val).(uint64))
-			paramType = pgtypes.Int64
-		case query.Type_YEAR:
-			paramType = pgtypes.Int16
-		case query.Type_FLOAT32:
-			paramType = pgtypes.Float32
-		case query.Type_FLOAT64:
-			paramType = pgtypes.Float64
-		case query.Type_DECIMAL:
-			paramType = pgtypes.Numeric
-		case query.Type_DATE, query.Type_DATETIME, query.Type_TIMESTAMP:
-			return nil, fmt.Errorf("need to add DoltgresType equivalents to DATETIME")
-		case query.Type_CHAR, query.Type_VARCHAR, query.Type_TEXT:
-			paramType = pgtypes.Text
-		case query.Type_ENUM:
-			paramType = pgtypes.Int16
-		case query.Type_SET:
-			paramType = pgtypes.Int64
-		case query.Type_NULL_TYPE:
-			paramType = pgtypes.Null
-		default:
-			return nil, fmt.Errorf("encountered an unknown GMS type")
-		}
-	}
-	castFunc := framework.GetImplicitCast(paramType.BaseID(), baseResultType.BaseID())
-	if castFunc == nil {
-		// This should never happen, but we'll check here just to be safe
-		resultType := array.getTargetType()
-		return nil, fmt.Errorf("cannot cast type %s to %s", resultType.BaseType().String(), paramType.String())
-	}
-	return castFunc(ctx, *val, baseResultType)
-}
-
-// isNullType returns whether the given type is a NULL type.
-func (array *Array) isNullType(t sql.Type) bool {
-	return t.Type() == query.Type_NULL_TYPE
-}
-
-// numberCastGroupPriority returns the priority for the given type belonging to the number group. The lower the value,
-// the higher the priority.
-func (array *Array) numberCastGroupPriority(t pgtypes.DoltgresTypeBaseID) int {
-	switch t {
-	case pgtypes.DoltgresTypeBaseID_Float64:
-		return 1
-	case pgtypes.DoltgresTypeBaseID_Float32:
-		return 2
-	case pgtypes.DoltgresTypeBaseID_Numeric:
-		return 3
-	case pgtypes.DoltgresTypeBaseID_Int64:
-		return 4
-	case pgtypes.DoltgresTypeBaseID_Int32:
-		return 5
-	case pgtypes.DoltgresTypeBaseID_Int16:
-		return 6
-	default:
-		return 7
-	}
-}
-
 // getTargetType returns the evaluated type for this expression. Returns the "anyarray" type if the type combination is
 // invalid.
 func (array *Array) getTargetType() pgtypes.DoltgresArrayType {
```

Comment:
We should process this in `WithChildren`, so that we're not calling this every time we call `Eval`. Types are static, so it's safe to do, and should make it a tad more performant, especially with the changes I mentioned to finding the common type. It's not entirely necessary for this PR, but if you can knock it out fairly quickly, it's worth doing, otherwise we should add a `TODO` for it.